File: DsAndAlgo/Python/src/Practice/patterns/sliding_window.py

def findLongestSubstring(s: str) -> int:
    if s is None:
        return -1
    if s == "":
        return 0
    if len(s) == 1:
        return 1

    last_index: dict[str, int] = {}
    start = 0
    max_len = 0

    for i, ch in enumerate(s):
        if ch in last_index and last_index[ch] >= start:
            start = last_index[ch] + 1
        last_index[ch] = i
        max_len = max(max_len, i - start + 1)

    return max_len

File: DsAndAlgo/GoLang/internal/patterns/slidingWindow.go
package patterns

func findLongestSubstring(str string) int {

	if str == "" {
		return -1
	}
	if len(str) == 0 {
		return 0
	}

	if len(str) == 1 {
		return 1
	}

	firstIdx := 0
	secondIdx := 0
	maxLen := 0

	set := make(map[byte]struct{})

	set[str[firstIdx]] = struct{}{}
	secondIdx = 1
	for secondIdx < len(str) {
		for _, ok := set[str[secondIdx]]; ok; _, ok = set[str[secondIdx]] {
			delete(set, str[firstIdx])
			firstIdx++
		}
		set[str[secondIdx]] = struct{}{}
		maxLen = max(maxLen, len(set))
		secondIdx++
	}

	maxLen = max(maxLen, len(set))

	return maxLen
}
